src/EditorFeatures/CSharpTest/ForEachCast/ForEachCastTests.cs

```diff
@@ -0,0 +1,1008 @@
+using Microsoft.CodeAnalysis.CSharp.Analyzers.ForEachCast;
+using Microsoft.CodeAnalysis.CSharp.CodeFixes.ForEachCast;
+using Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ForEachCast
+{
+    using VerifyCS = CSharpCodeFixVerifier<
+        CSharpForEachCastDiagnosticAnalyzer,
+        CSharpForEachCastCodeFixProvider>;
+
+    public class ForEachCastTests
+    {
+        private static async Task TestWorkerAsync(
+            string testCode, string fixedCode, string optionValue)
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = testCode,
+                FixedCode = fixedCode,
+                EditorConfig = @"
+[*]
+dotnet_style_prefer_foreach_explicit_cast_in_source=" + optionValue,
+            }.RunAsync();
+        }
+
+        private static Task TestAlwaysAsync(string markup, string alwaysMarkup)
+            => TestWorkerAsync(markup, alwaysMarkup, "always");
+
+        private static Task TestWhenStronglyTypedAsync(string markup, string nonLegacyMarkup)
+            => TestWorkerAsync(markup, nonLegacyMarkup, "when_strongly_typed");
+
+        [Fact]
+        public async Task NonGenericIComparableCollection()
+        {
+            var test = @"
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            [|foreach|] (string item in new A())
+            {
+            }
+        }
+    }
+    struct A
+    {
+        public Enumerator GetEnumerator() =>  new Enumerator();
+        public struct Enumerator
+        {
+            public System.IComparable Current => 42;
+            public bool MoveNext() => true;
+        }
+    }
+}";
+
+            await TestAlwaysAsync(test, test);
+            await TestWhenStronglyTypedAsync(test, test);
+        }
+
+        [Fact]
+        public async Task GenericObjectCollection()
+        {
+            var test = @"
+using System.Collections.Generic;
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            var x = new List<object>();
+            [|foreach|] (string item in x)
+            {
+            }
+        }
+    }
+}";
+            var fixedCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            var x = new List<object>();
+            foreach (string item in x.Cast<string>())
+            {
+            }
+        }
+    }
+}";
+
+            await TestAlwaysAsync(test, fixedCode);
+            await TestWhenStronglyTypedAsync(test, fixedCode);
+        }
+
+        [Fact]
+        public async Task ObjectArray()
+        {
+            var test = @"
+using System.Collections.Generic;
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main(object[] x)
+        {
+            [|foreach|] (string item in x)
```

Comment:
I'm curious about testing tuples

```csharp
        void Main((object, object)[] x)
        {
            foreach ((string, string) item in x)
            {
            }
        }
```

Reply:
This is interesting due to:

```csharp
using System;
using System.Linq;

Program.M(new (object, object)[] { ("AB", "CD") });

public static class Program
{   
    public static void M((object, object)[] x)
    {
        foreach ((string, string) item in x
            //.Cast<(string, string)>() // Uncomment this. Crash!!
        )
        {
            Console.WriteLine(item.Item1);
            Console.WriteLine(item.Item2);
        }
    }
}
```

Reply:
@CyrusNajmabadi I found them, but none seems to match the above case.

Reply:
> //.Cast<(string, string)>() // Uncomment this. Crash!!

Added test to show we do the right thing here (no crash) :)